smoketests/__init__.py

```diff
@@ -0,0 +1,195 @@
+from pathlib import Path
+import unittest
+import tempfile
+import os
+import re
+import string
+import random
+import string
+import subprocess
+import json
+import sys
+import shutil
+
+TEST_DIR = Path(__file__).parent
+STDB_DIR = TEST_DIR.parent
+SPACETIME_BIN = STDB_DIR / "target/debug/spacetime"
+TEMPLATE_TARGET_DIR = STDB_DIR / "target/_stdbsmoketests"
+STDB_CONFIG = TEST_DIR / "config.toml"
+
+TEMPLATE_LIB_RS = open(STDB_DIR / "crates/cli/src/subcommands/project/rust/lib._rs").read()
+TEMPLATE_CARGO_TOML = open(STDB_DIR / "crates/cli/src/subcommands/project/rust/Cargo._toml").read()
+bindings_path = (STDB_DIR / "crates/bindings").absolute()
+TEMPLATE_CARGO_TOML = (re.compile(r"^spacetimedb\s*=.*$", re.M) \
+    .sub(f'spacetimedb = {{ path = "{bindings_path}" }}', TEMPLATE_CARGO_TOML))
+
+HAVE_DOCKER = False
+
+
+def build_template_target():
+    if not TEMPLATE_TARGET_DIR.exists():
+        print("Building base compilation artifacts")
+        class BuildModule(Smoketest):
+            AUTOPUBLISH = False
+
+        BuildModule.setUpClass()
+        env = { **os.environ, "CARGO_TARGET_DIR": TEMPLATE_TARGET_DIR }
+        spacetime("build", "-Sd", BuildModule.project_path, env=env, capture_stderr=False)
+        BuildModule.tearDownClass()
+        BuildModule.doClassCleanups()
+
+
+def requires_docker(item):
+    if HAVE_DOCKER:
+        return item
+    return unittest.skip("docker not available")(item)
+
+def random_string(k=20):
+    return ''.join(random.choices(string.ascii_letters, k=k))
+
+def extract_fields(cmd_output, field_name):
```

Comment:
I think is fine to assume we have control over the `parsing` input, but maybe add here some docs showing how the data looks like?